pandas/core/arrays/categorical.py

```diff
@@ -2717,6 +2718,15 @@ def _groupby_op(
             res_values[result_mask == 1] = -1
         return self._from_backing_data(res_values)
 
+    def _factorize_with_other(
+        self, other: Categorical, sort: bool = False  # type: ignore[override]
+    ) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.intp], int]:
+        other = self._encode_with_my_categories(other)
```

Comment:
i think that factorize_with_other assumes matching dtypes. if correct, comment/docstring/assert to that effect?